### PR TITLE
QB: evaluate notebook preview lazy

### PR DESCRIPTION
**Pull request by uladzimirdev**
### Description

Calculate query preview in a lazy way

### How to verify

- open notebook editor
- add any combination of filters, aggregations and grouping
- click on preview button
- make sure preview is calculated correctly in all cases

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/notebook/NotebookStep/NotebookStep.tsx [hunk 1]
```diff
@@ -104,7 +104,7 @@ function NotebookStep({
   } = STEP_UI[step.type] || {};
 
   const color = getColor();
-  const canPreview = Boolean(step.previewQuery);
+  const canPreview = Boolean(step.getPreviewQuery);
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/notebook/NotebookStep/NotebookStep.tsx:107]: can be `step.active && step.visible` instead

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.ts [hunk 2]
```diff
@@ -304,15 +302,36 @@ function getStageSteps(
...
+  stepIndex: number,
+): Lib.Query {
+  const previewQuery = _.range(
+    Lib.stageCount(query) - 1,
+    stageIndex,
+    -1,
+  ).reduce(Lib.dropStage, query);
+
+  // revert query for this step
+  const queryAfterRevertingSteps = _.range(
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/notebook/lib/steps.ts:322]: nit:

```
steps
  .slice(stepIndex, steps.length)
  .reverse()
  .reduce((query, step) => ...)
````

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.ts [hunk 3]
```diff
@@ -304,15 +302,36 @@ function getStageSteps(
...
 
   return { steps, actions };
 }
+function getPreviewQuery(
+  query: Lib.Query,
+  stageIndex: number,
+  steps: NotebookStep[],
+  stepIndex: number,
+): Lib.Query {
+  const previewQuery = _.range(
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/notebook/lib/steps.ts:315]: nit: maybe name both like `queryWithoutNextStages` and `queryWithoutNextSteps` or something like this

**Review by ranquild — APPROVED** [review]

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.ts (hunk 3 above)
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/notebook/lib/steps.ts]: used `queryWithoutNextStages` and `queryWithoutNextStagesAndSteps`

Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/notebook/lib/steps.ts]: 93a896bf640e4921751a9418ba2952bc023aecb5

cleaner, indeed

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.ts (hunk 2 above)
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/notebook/lib/steps.ts]: 93a896bf640e4921751a9418ba2952bc023aecb5

cleaner, indeed

**Pull request merged**